### PR TITLE
bpo-35066: _dateime.datetime.strftime copies trailing '%'

### PR DESCRIPTION
**Pull request by michaelsaah**
https://bugs.python.org/issue35066

https://bugs.python.org/issue35066

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by MichaelSaah** [comment]
Summary (also posted to tracker):

Modules/_datetimemodule.c and Lib/datetime.py do not behave identically. Specifically, the strftime functions do not match when passed a format string
terminated with a '%'. The C function performs an explicit check for this condition, and raises a ValueError on it. The Py version does not perform this check. Both pass the
format string (after doing substitutions for %z, %Z, and %f tags) to the system strftime or wcfstime, depending on platform. These live within the python time module. The
time module wrapper function does not perform this check.

This situation leads to a scenario in which, for example, "%D %" passed to datetime.strftime (with the C extension included) raises a value error. The same string passed to
time.strftime returns "mm/dd/yy %", at least on OSX. Furthermore, if Python is built without the C module, "mm/dd/yy %" is returned when datetime.strftime is called.

To summarise, there are two problems: (1) datetime does not comply with PEP-399, and (2) a higher-order module raises an exception on a case that the (exposed) lower-order
module has no problem with, causing a mismatch in behavior between datetime.strftime and time.strftime.

This PR attempts to fix this problem by removing the case check from the datetime C module. This solves both (1) and (2).

There was much talk on the issue thread about there existing a test case for time.strftime that documented a platform-dependent failure on a dangling '%'. I wish to note
that my patch does not touch the time module at all, it only removes a seemingly unnecessary check in the datetime C module.

**Comment by pganssle** [comment]
@MichaelSaah @tirkarthi Can either of you point to the existing test case on this? I cannot find it.

I think that this still needs a new test, because obviously the existing test case does not cover it specifically. Ideally you would want to test that this does not fail for `datetime.strftime` specifically. If there's a platform-dependent failure in `time.strftime`, then you can write the test such that the `datetime` test is skipped if the `time` test would fail, e.g.:

```python
# Whether datetime.strftime can handle a trailing % is platform-dependent,
# so detect whether we're on one of the platforms where this fails (bpo-35066)
try:
    time.strftime('%')
    skip_trailing_percent_strftime = False
except ValueError:
    skip_trailing_percent_strftime = True
```

This is assuming that we don't already *know* the platforms where the trailing-% fails. If we know which platforms this fails on, then both the `time.strftime` and `datetime.strftime` tests should be skipped on the platform with the known failure.

**Comment by tirkarthi** [comment]
@pganssle In the tracker I was referring to https://github.com/python/cpython/blob/9e95eb0d609cee23e6c9915c0bef243585b8c14b/Lib/test/test_time.py#L240 which was introduced with e5b5895b5b5 and has an explicit comment about the test being platform dependent.

**Comment by pganssle** [comment]
@tirkarthi Hm. That is a very unfortunately designed test. Even legitimate failures will be swallowed because it treats *all* `ValueError`s as success, even if there is no known platform where this is a failure case.

I think for now we should try the "detect if `time.strftime` fails, otherwise test `datetime.strftime`" method, and maybe in the future we can tighten up the failure conditions for `time.strftime`.

**Comment by MichaelSaah** [comment]
@pganssle Thanks Paul. I'm not sure if the platform failure modes are known; I didn't run into documentation of them during my work. I'll work on adding a test case for `datetime` that is aware of `time.strftime`'s functionality, but it would help me to know what exactly would happen if stdlib call in `time.strftime` failed, so that I can catch the failure cleanly. Otherwise I guess I'll be spinning up a bunch of VMs and trying to induce failure.

It looks like there's some error handling being done by `_Py_BEGIN_SUPPRESS_IPH`, but I'm not sure how this propagates (`format_time` is a macro for `strftime`/`wcfstime` depending on platform):

https://github.com/python/cpython/blob/163eca34c48f1b25e1504e37f4656773fd0fdc78/Modules/timemodule.c#L824-L826

The macro def'ns are a bit cryptic:

https://github.com/python/cpython/blob/542497aa9f71c664768c3d5b7398c03679d3a7e1/Include/pyport.h#L763-L773

Any help appreciated.

**Comment by pganssle** [comment]
@MichaelSaah I don't think you have to worry about any of that. We know that `time.strftime('%')` should *either* raise `ValueError` *or* succeed. You can just write a test for `datetime.strftime` that assumes that `datetime_instance.strftime('%')` returns `'%'`, and then skip that test conditional on the logic I described in my earlier comment (i.e. skip it on all platforms where `time.strftime('%')` raises `ValueError`).

**Comment by MichaelSaah** [comment]
@pganssle Ah ok, I was reading the comment in the test-in-question incorrectly. I see the spec now. Thanks.

**Comment by MichaelSaah** [comment]
Just added the test. Will probably open another PR to try to cleanup the datetime.strftime tests, they're all over the place.

Interestingly, looks like there's been some confusion about this before:
https://github.com/python/cpython/blob/65c216e74f7957006ef7653b7e2afe83007c45ce/Lib/test/datetimetester.py#L1330-L1342

File: Lib/test/datetimetester.py [hunk 1]
```diff
@@ -1350,6 +1350,17 @@ def test_strftime(self):
         #check that this standard extension works
         t.strftime("%f")
 
+        # bpo-35066: make sure trailing '%' don't cause
+        # datetime's strftime to complain
+        failed = False
+        try:
+            t.strftime('%')
```
Comment (pganssle) [inline Lib/test/datetimetester.py]:
```suggestion
            _time.strftime('%')
```

I believe the way you currently have it would not fail before your modifications.  What you've done is basically a long-winded version of:

```python
try:
    self.assertEqual(t.strftime('%'), '%')
    self.assertEqual(t.strftime('m:%m d:%d y:%y %'), 'm:03 d:02 y:05 %')
except ValueError:
    pass
```

What you *actually* want to check is that `datetime.datetime.strftime` doesn't fail *if trailing % signs are supported in `time.strftime`*.

File: Lib/test/datetimetester.py [hunk 2]
```diff
@@ -1350,6 +1350,17 @@ def test_strftime(self):
         #check that this standard extension works
         t.strftime("%f")
 
+        # bpo-35066: make sure trailing '%' don't cause
+        # datetime's strftime to complain
+        failed = False
```
Comment (pganssle) [inline Lib/test/datetimetester.py]: Can you rename this to something more descriptive? The way it is now it's not clear what is failing. I would probably do it like this:

```python
time_supports_trailing_percent = True
try:
    _time.strftime('%')
except ValueError:
    time_supports_trailing_percent = False
```

File: Lib/test/datetimetester.py [hunk 3]
```diff
@@ -1350,6 +1350,17 @@ def test_strftime(self):
...
+        try:
+            t.strftime('%')
+        except ValueError:
+            failed = True
+        if not failed:
```
Comment (pganssle) [inline Lib/test/datetimetester.py]: I would prefer this to be in a subtest, like so:

```python
with self.subTest('trailing percentage'):
    if not time_supports_trailing_percent:
        self.skipTest('time module does not support trailing %')

    self.assertEqual(t.strftime('%'), '%'))
    self.assertEqual(t.strftime("m:%m d:%d y:%y %"), "m:03 d:02 y:05 %")
```

Though I'm thinking maybe this should instead be moved to its own test method. Partially because I tend to prefer a larger number of small, targeted test methods rather than big mega-tests and partially because `skipTest` and `subTest` when used together apparently give a misleading message. An example:

```python
import unittest

class SomeTest(unittest.TestCase):
    def test_something(self):
        for i in range(1, 3):
            with self.subTest(i):
                if i > 1:
                    self.skipTest('Not supported')

                self.assertEqual(i, 1)
```

When you run this with `python -m unittest -v`, you get:

```
$ python -m unittest -v
test_something (test_mod.SomeTest) ... skipped 'Not supported'

----------------------------------------------------------------------
Ran 1 test in 0.000s

OK (skipped=1)
```

And with `pytest` you get:

```
========= test session starts =======

platform linux -- Python 3.7.1, pytest-4.0.1, py-1.7.0, pluggy-0.8.0
rootdir: /tmp/test_mod, inifile:
collected 1 item

test_mod.py s                                           [100%]

===== 1 skipped in 0.00 seconds =====
```

Both are misleading, because only the `i == 2` subtest is being skipped *not* the whole test.

**Review by pganssle — CHANGES_REQUESTED** [review]

File: Lib/test/datetimetester.py (hunk 1 above)
Comment (michaelsaah) [reply Lib/test/datetimetester.py]: Good catch, I intended to call the `time` module, not use the date object's wrapper method.

File: Lib/test/datetimetester.py (hunk 2 above)
Comment (michaelsaah) [reply Lib/test/datetimetester.py]: Ok, will do

File: Lib/test/datetimetester.py (hunk 3 above)
Comment (michaelsaah) [reply Lib/test/datetimetester.py]: Ok, so parking the test in it's own method will allow me to skip it on a failing `time` call without making it look like all the other tests in that method were skipped.

File: Lib/test/datetimetester.py [hunk 4]
```diff
@@ -1350,6 +1350,17 @@ def test_strftime(self):
         #check that this standard extension works
         t.strftime("%f")
 
+    def test_strftime_trailing_percent(self):
+    # bpo-35066: make sure trailing '%' doesn't cause
```
Comment (pganssle) [inline Lib/test/datetimetester.py]: Can you indent these comments so they are in line with the function body?

File: Modules/_datetimemodule.c [hunk 5]
```diff
@@ -1521,20 +1521,15 @@ wrap_strftime(PyObject *object, PyObject *format, PyObject *timetuple,
     if (newfmt == NULL) goto Done;
     pnew = PyBytes_AsString(newfmt);
     usednew = 0;
+    ch = *pin;
 
-    while ((ch = *pin++) != '\0') {
-        if (ch != '%') {
+    do {
```
Comment (pganssle) [inline Modules/_datetimemodule.c]: Hm... It's a bit hard to keep track of where null checking happens in this loop.

I guess the reason that trailing `%` was forbidden in the original version was that they wanted to check for null termination every time the `pin` pointer is advanced.

I think there's a change in behavior around null-checking here, because before this change, the new format string did not include the null terminator, after the change, the new format string does include the null terminator. I don't know that this is a *problem*, but I'm wary of it.

Maybe you can change it like this:

```c
while((ch = *pin++) != '\0') {
    if (ch != '%' || (ch = *pin++) == '\0') {
        ptoappend = pin - 1;
        ntoappend = 1;
    }
    else if (ch == 'z') {
    }
```

That should keep the logic the same.

Comment (pganssle) [reply Modules/_datetimemodule.c]: As an aside, this terse C-style stuff where you have to keep track of null pointers is for the birds. It took me a long time to convince myself that there was a change in behavior, and even still I am not certain about it. It really feels weird to advance the pointer *in the conditionals*.

Comment (michaelsaah) [reply Modules/_datetimemodule.c]: Just looking at it, I don't _think_ your code does what you think it does. Without a null char check at the end of the loop, you'll read the null char and advance the pointer, and then the check in the while loop condition will actually be reading the char _after_ the null char.

Like you said though, this stuff is tricky, so I'm going to test it and report back.

Also, my first attempt at fixing this, while I think a bit uglier than the do/while solution, comes closer to your desired behavior then the current one. Here it is:

https://github.com/python/cpython/blob/d178065d5a4e76c00bb1a14dd3d20e935f913537/Modules/_datetimemodule.c#L1525-L1626

As far as the pointer juggling in the conditionals goes, it's a pattern I've seen before in C parsing code. I think ideally you only mutate the pointer in one place, as it quickly becomes hard to reason about.

Comment (michaelsaah) [reply Modules/_datetimemodule.c]: As I suspected, your code doesn't do what we want. Passing it "%D %" produces the format string `b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb'`.

I obtained this by calling `printf("%s\n", PyUnicode_AsUTF8(PyObject_Repr(newfmt)));` after the while loop. If you put the call in the loop itself, you can watch it happen: 
```
b'%D\xcb\xcb\xcb'
b'%D \xcb\xcb'
b'%D \x00\xcb'
b'%D \x00\xfb'
b'%D \x00\xfb\xfb\xcb\xcb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xcb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb'
b'%D \x00\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xfb\xcb\xcb\xcb\xcb\xcb\xcb\xcb\xcb'
```

I'm not sure what's causing it to eventually break out of the loop and not spin forever; I assume it's one of the length checks that happens in the loop.

File: Lib/test/datetimetester.py (hunk 4 above)
Comment (michaelsaah) [reply Lib/test/datetimetester.py]: Done

File: Modules/_datetimemodule.c (hunk 5 above)
Comment (pganssle) [reply Modules/_datetimemodule.c]: @MichaelSaah Good call, sorry, I didn't actually try my version, you are right that I advance but then never check if the next character is a `NULL`. I think maybe this will do what I was looking for?

```c
while((ch = *pin++) != '\0') {
    if (ch != '%' || *pin == '\0') {
        ptoappend = pin - 1;
        ntoappend = 1;
    }
    else if ((ch = *pin++) == 'z') {
    }
```

**Edit**: No, just realized why this won't work. Hm... If I'm right that the behavior is slightly different, we may need to have more than one check in here, unfortunately.

**Comment by tirkarthi** [comment]
@MichaelSaah You might want to add a NEWS entry given the impact and it will also help in keeping the PR green : https://devguide.python.org/committing/#what-s-new-and-news-entries

File: Modules/_datetimemodule.c (hunk 5 above)
Comment (michaelsaah) [reply Modules/_datetimemodule.c]: Did you check the earlier commit I linked to? That might be a blueprint for what you want. The key is the null check and break at the end of the loop.

**Comment by MichaelSaah** [comment]
@pganssle I think you'll like the latest change. Behavior w/ regard to null bytes now matches existing code, as far as I can tell.

**Comment by pganssle** [comment]
@MichaelSaah Sorry for the days of silence.

I think the *behavior* of this is now fixed. I'm not sure how I feel about decrementing the pointer, though. I think the two most obviously viable solutions (barring any that dramatically refactor the code, which I have not considered):

1. Checking for a null at the beginning and the end of the loop
2. Checking for a null at the beginning of the loop, but decrementing the pointer (as you've done in this case) when you find a second null

The first one is nice because it doesn't go through unnecessary iterations of the loop, but the second one keeps the null checking logic as close as possible to the point where the pointer is incremented, which keeps the amount of time between reaching the end of the string and knowing we've reached it to a minimum.

I think the only other decent option would be something like this:

```c
while ((ch = *pin++) != '\0') {
    if(ch != '%') {
        ptoappend = pin - 1;
        ntoappend = 1;
    } else if (*(pin + 1) != '\0') {
        if ((ch = *pin++) == 'z') {
            ...
```

I am increasingly convincing myself that number 2 (your current version) is the right way to go, so I say we go ahead and merge this if no one else objects.

**Review by pganssle — APPROVED** [review]

**Comment by pganssle** [comment]
By the way, thank you for your patience with this @MichaelSaah, particularly since I know I created some extra work for you with my ill-considered alternative approaches.

**Comment by MichaelSaah** [comment]
Wow we really commented past each other there. No worries about the extra work, you've been a very clear and thorough reviewer, and I'm doing this for fun while looking for a new job anyway.

I originally didn't want to decrement the pointer either, but I'm convinced it's safe, since we only do that if we've already seen a '%', which means we won't be decrementing past the beginning of the buffer.

After staring at this for hours over the past couple weeks, I'm not convinced that there's any way to make the control flow more intuitive while keeping the null-checking contained.

As far as the difference between (1) and (2), I think (2) fits more naturally with the case-checking semantics that are already there, but it's a toss-up. I don't think either is more confusing (or clearer) than the other.

**Comment by MichaelSaah** [comment]
As an aside, I don't see how my solution leads to unnecessary traversals of the loop. As soon as it sees a '%' followed by a '\0', it copies the '%' and breaks on the next while condition check. Do you mean that it breaks on the while condition instead of at the bottom of the loop body?

**Comment by pganssle** [comment]
CC: @abalkin

I think this is ready to be merged.
